src/main/scala/ch/epfl/bluebrain/nexus/admin/index/DistributedDataIndex.scala

```diff
@@ -0,0 +1,188 @@
+  }
+
+  override def getOrganization(label: String): F[Option[ResourceF[Organization]]] =
+    fetchFromMap(labelToOrganizationsKey, label)
+
+  override def getProject(organization: String, project: String): F[Option[ResourceF[Project]]] =
+    fetchFromMap(labelToProjectsKey, s"$organization/$project")
+
+  override def listOrganizations(pagination: Pagination): F[List[ResourceF[Organization]]] =
+    fetchList(organizationsListKey, pagination)
+
+  override def listProjects(pagination: Pagination): F[List[ResourceF[Project]]] =
+    fetchList(projectsListKey, pagination)
+
+  override def listProjects(organizationLabel: String, pagination: Pagination): F[List[ResourceF[Project]]] = {
+    getOrganization(organizationLabel).flatMap {
+      case None => F.pure(List.empty)
+      case Some(org) =>
+        fetchList(projectsOrgListKey(org.uuid), pagination)
+    }
+  }
+
+  private def fetchList[A](dDataKey: LWWRegisterKey[RevisionedValue[SortedSet[ResourceF[A]]]],
+                           pagination: Pagination): F[List[ResourceF[A]]] = {
+    val get = Get(dDataKey, ReadMajority(consistencyTimeout), None)
+    IO.fromFuture(IO(replicator ? get)).to[F].flatMap {
+      case g @ GetSuccess(`dDataKey`, _) =>
+        F.pure(
+          g.get(dDataKey)
+            .value
+            .value
+            .slice(pagination.from.toInt, (pagination.from + pagination.size).toInt)
+            .toList)
+      case NotFound(_, _) => F.pure(List.empty)
+    }
+  }
+
+  private def updateUuidToResource[A: Clock](dDataKey: LWWMapKey[String, A], mapKey: String, value: A): F[Unit] = {
+    val msg =
+      Update(dDataKey, LWWMap.empty[String, A], WriteMajority(consistencyTimeout)) {
+        _.put(mapKey, value)
+      }
+    update(msg, s"update $mapKey -> resource")
+  }
+
+  private def updateList[A](dDataKey: LWWRegisterKey[RevisionedValue[SortedSet[ResourceF[A]]]], value: ResourceF[A])(
+      implicit ordering: Ordering[ResourceF[A]]): F[Unit] = {
+    val msg = Update(
+      dDataKey,
+      LWWRegister(RevisionedValue.apply[SortedSet[ResourceF[A]]](0, SortedSet.empty[ResourceF[A]])),
+      WriteMajority(consistencyTimeout)
+    )(updateWithIncrement(_, value))
+    update(msg, s"update ${value.uuid.toString} in ${dDataKey._id}")
+  }
+
+  private def fetchFromMap[A](dDataKey: LWWMapKey[String, A], mapKey: String): F[Option[A]] = {
+    val get = Get(dDataKey, ReadMajority(consistencyTimeout), None)
+    IO.fromFuture(IO(replicator ? get)).to[F].flatMap {
+      case g @ GetSuccess(`dDataKey`, _) =>
+        F.pure(g.get(dDataKey).get(mapKey))
+      case NotFound(_, _) => F.pure(None)
+    }
+  }
+
+  private def update(update: Update[_], action: String): F[Unit] = {
+    IO.fromFuture(IO(replicator ? update)).to[F].flatMap {
+      case UpdateSuccess(_, _) =>
+        F.unit
+      case UpdateTimeout(LWWRegisterKey(_), _) =>
+        F.raiseError(new RetriableErr(s"Distributed cache update timed out while performing action '$action'"))
```

Comment:
I'm not fond of these retryable errors; I think the cache should take a `RetryStrategy[F]` (the type from sourcing) and wrap the execution with those strategies like. Here's an example of what I mean for an arbitrary put:
```
def put(key: K, value: V): F[Unit] = {
  val msg    = Update(mapKey, LWWMap.empty[K, V], WriteAll(consistencyTimeout))(_.put(key, value))
  val future = IO(replicator ? msg)
  val fa     = IO.fromFuture(future).to[F]
  val mappedFA = fa.flatMap[Unit] {
    case _: UpdateSuccess[_] => F.unit
    // $COVERAGE-OFF$
    case _: UpdateTimeout[_] => F.raiseError(InternalError(consistencyTimeoutError))
    case _: UpdateFailure[_] => F.raiseError(InternalError(DistributedDataError("Failed to distribute write")))
    // $COVERAGE-ON$
  }
  retryStrategy(mappedFA)
}
```

Reply:
The retry is already handled in the `SequentialTagIndexer`, we should probably unify `sourcing and `indexing` to use the same kind of `RetryStrategy`

Reply:
Agreed.